Apps/Src/Tasks.c

```diff
@@ -45,28 +46,110 @@ CPU_STK DebugDump_Stk[TASK_DEBUG_DUMP_STACK_SIZE];
 CPU_STK CommandLine_Stk[TASK_COMMAND_LINE_STACK_SIZE];
 
 /**
- * Semaphores
+ * Global Variables
  */
-OS_SEM FaultState_Sem4;
+os_error_loc_t OSErrLocBitmap = OS_NONE_LOC; // Store the location of the current error
+
+// Variables to store error codes, stored and cleared in task error assert functions
+error_code_t Error_ReadCarCAN = /*READCARCAN_ERR_NONE*/ 0; // TODO: change this back to the error 
+error_code_t Error_ReadTritium = T_NONE;  // Initialized to no error
+error_code_t Error_UpdateDisplay = UPDATEDISPLAY_ERR_NONE;
+
+extern const pinInfo_t PININFO_LUT[]; // For GPIO writes. Externed from Minions Driver C file.
 
 /**
- * Global Variables
- */
-fault_bitmap_t FaultBitmap = FAULT_NONE;
-os_error_loc_t OSErrLocBitmap = OS_NONE_LOC;
+ * Error assertion-related functions
+*/
 
-void _assertOSError(uint16_t OS_err_loc, OS_ERR err)
+void _assertOSError(os_error_loc_t OS_err_loc, OS_ERR err)
 {
     if (err != OS_ERR_NONE)
     {
-        FaultBitmap |= FAULT_OS;
-        OSErrLocBitmap |= OS_err_loc;
+        EmergencyContactorOpen(); // Turn off contactors and turn on the brakelight to indicate an emergency
+        Display_Error(OS_err_loc, err); // Display the location and error code
+        while(1){;} //nonrecoverable
 
-        OSSemPost(&FaultState_Sem4, OS_OPT_POST_1, &err);
-        EnterFaultState();
     }
 }
 
+/**
+ * @brief Assert a task error by locking the scheduler (if necessary), displaying a fault screen,
+ * and jumping to the error's specified callback function. 
+ * Called by task-specific error-assertion functions that are also responsible for setting the error variable.
+ * @param errorLoc the task from which the error originated. Note: should be taken out when last task pointer is integrated
+ * @param errorCode the enum for the specific error that happened
+ * @param errorCallback a callback function to a handler for that specific error, 
+ * @param lockSched whether or not to lock the scheduler to ensure the error is handled immediately. Only applicable for recoverable errors- nonrecoverable errors will always lock
+ * @param nonrecoverable whether or not to kill the motor, display the fault screen, and enter an infinite while loop
+*/
+void assertTaskError(os_error_loc_t errorLoc, error_code_t errorCode, callback_t errorCallback, error_scheduler_lock_opt_t lockSched, error_recov_opt_t nonrecoverable) {
```

Comment:
Ah I see- I'll change it to throwTaskError instead. 

I'll also add a check for errorCode == 0. The way the error function is set up, the assertError functions for individual tasks should check the errorCode and only call this function if errorCode != 0. However, given that all our error enums use 0 to indicate no error anyways, it's probably safe to include an additional check here just in case.